packages/discovery-provider/plugins/pedalboard/apps/solana-relay/src/routes/relay/relay.ts

```diff
@@ -0,0 +1,169 @@
+      kp.publicKey.equals(feePayerPublicKey)
+    ) ?? null
+  )
+}
+
+const forwardTransaction = async (logger: Logger, transaction: string) => {
+  const endpoints = await getCachedDiscoveryNodeEndpoints()
+  logger.info(`Forwarding to ${endpoints.length} endpoints...`)
+  const body = JSON.stringify({ transaction })
+  await Promise.all(
+    endpoints
+      .filter((endpoint) => endpoint !== config.endpoint)
+      .map((endpoint) =>
+        fetch(`${endpoint}/solana/cache`, {
+          method: 'POST',
+          body,
+          headers: {
+            'content-type': 'application/json',
+            'Discovery-Signature': personalSign(config.delegatePrivateKey, {
+              data: body
+            })
+          }
+        })
+          .then((res) => {
+            if (res.ok) {
+              logger.info(
+                { endpoint, status: res.status },
+                `Forwarded successfully`
+              )
+            } else {
+              logger.warn(
+                { endpoint },
+                `Failed to forward transaction to endpoint: ${res.statusText}`
+              )
+            }
+          })
+          .catch((e) => {
+            logger.warn(
+              { endpoint },
+              `Failed to forward transaction to endpoint: ${e}`
+            )
+          })
+      )
+  )
+}
+
+export const relay = async (
+  req: Request<unknown, unknown, RequestBody>,
+  res: Response,
+  next: NextFunction
+) => {
+  try {
+    const { transaction: encodedTransaction, confirmationOptions } = req.body
+    const { strategy, commitment } = confirmationOptions ?? {}
+    const confirmationStrategy =
+      strategy ?? (await connection.getLatestBlockhash())
+    const decoded = Buffer.from(encodedTransaction, 'base64')
+    const transaction = VersionedTransaction.deserialize(decoded)
+    const decompiled = TransactionMessage.decompile(transaction.message)
+    const feePayerKey = transaction.message.getAccountKeys().get(0)
+    const feePayerKeyPair = getFeePayerKeyPair(feePayerKey)
+    if (!feePayerKey || !feePayerKeyPair) {
+      throw new BadRequestError(
+        `No fee payer for address '${feePayerKey?.toBase58()}'`
+      )
+    }
+    await assertRelayAllowedInstructions(decompiled.instructions, {
+      user: res.locals.signerUser,
+      feePayer: feePayerKey.toBase58()
+    })
+
+    transaction.sign([feePayerKeyPair])
+
+    const logger = res.locals.logger.child({
+      signature: base58.encode(transaction.signatures[0])
+    })
+    logger.info('Sending transaction...')
+    const serializedTx = transaction.serialize()
+
+    const signature = await connection.sendRawTransaction(serializedTx)
+    if (commitment) {
+      logger.info(`Waiting for transaction to be ${commitment}...`)
+      await connection.confirmTransaction(
+        {
+          ...confirmationStrategy,
+          signature
+        },
+        commitment
+      )
+    }
+    res.status(200).send({ signature })
+    next()
```

Comment:
i'm not familiar with `next()`, seems like it's for yielding the next value in a generator, but don't think that's the case here?

Reply:
basically i'm not clear on how the control flow for the rest of this `relay` function works from here on out.

Reply:
basically this lets the rest of the middleware run and sends the response, but we can still do more work after. It's very unconventional - usually sending the response is the last thing you do.